examples/lang/shared_vec/src/lib.rs

```diff
@@ -0,0 +1,252 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// ink! is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with ink!.  If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(any(test, feature = "test-env")), no_std)]
+
+use ink_core::{
+    env::AccountId,
+    storage,
+};
+use ink_lang::contract;
+use parity_codec::{Encode, Decode};
+
+/// Access rights to the shared vector.
+#[derive(Encode, Decode)]
+pub struct Access {
+    pub begin: Option<u32>,
+    pub end: Option<u32>,
+}
+
+impl Access {
+    /// Creates unlimited access rights.
+    pub fn unlimited() -> Self {
+        Access { begin: None, end: None }
+    }
+
+    /// Returns `true` if the access is unlimited.
+    pub fn is_unlimited(&self) -> bool {
+        self.begin.is_none() && self.end.is_none()
+    }
+
+    /// Creates new limited access rights.
+    pub fn new<B, E>(begin: B, end: E) -> Self
+    where
+        B: Into<Option<u32>>,
+        E: Into<Option<u32>>,
+    {
+        let begin = begin.into();
+        let end = end.into();
+        assert!(begin <= end);
+        Access { begin, end }
+    }
+
+    /// Returns `true` if the given index is within access rights.
+    pub fn contains(&self, index: u32) -> bool {
+        let begin = self.begin.unwrap_or(0);
+        let end = self.end.unwrap_or(core::u32::MAX);
+        begin <= index && index <= end
+    }
+}
+
+pub type ErrNo = u32;
+pub type Result<T, E> = core::result::Result<T, E>;
+
+const ACCESS_NOT_ALLOWED: ErrNo = 1;
+const NOT_REGISTERED: ErrNo = 2;
+const OUT_OF_BOUNDS: ErrNo = 3;
+
+contract! {
+    /// A shared vector that is accessiable to a subset of allowed mutators.
+    struct SharedVec {
+        /// The allowed mutators.
+        ///
+        /// They can operate on a range within the shared vector.
+        /// The range is defined by `(start,end)` where `start` and `end`
+        /// refer to the zero-starting index. A value of `None` means
+        /// that there is no lower or upper bound.
+        mutators: storage::HashMap<AccountId, Access>,
+        /// The shared vector.
+        vec: storage::Vec<i32>,
+    }
+
+    /// Fires whenever a new mutator is registered
+    /// or when a mutators access rights are changed.
+    event Register {
+        // / The mutator.
+        mutator: AccountId,
+        // / The begin access index.
+        begin: Option<u32>,
+        // / The end access index.
+        end: Option<u32>,
+    }
+
+    /// Fires whenever a mutator pushes the vector successfully.
+    event Push {
+        // / The pushed value.
```

Comment:
It makes sense to me to include the `mutator: AccountId` with these events... not for a practical reason, but more like an "imaginary UI/UX" reason

Reply:
good idea, will do that!